lib/client/presence/remote-doc-presence.js

```diff
@@ -0,0 +1,151 @@
+var RemotePresence = require('./remote-presence');
+var ot = require('../../ot');
+
+module.exports = RemoteDocPresence;
+function RemoteDocPresence(presence, presenceId) {
+  RemotePresence.call(this, presence, presenceId);
+
+  this.collection = this.presence.collection;
+  this.id = this.presence.id;
+  this.src = null;
+
+  this._doc = this.connection.get(this.collection, this.id);
+  this._pending = null;
+  this._opCache = null;
+  this._pendingSetPending = false;
+
+  this._opHandler = this._handleOp.bind(this);
+  this._createDelHandler = this._handleCreateDel.bind(this);
+  this._loadHandler = this._handleLoad.bind(this);
+  this._registerWithDoc();
+}
+
+RemoteDocPresence.prototype = Object.create(RemotePresence.prototype);
+
+RemoteDocPresence.prototype.receiveUpdate = function(message) {
+  if (this._pending && message.seq < this._pending.seq) return;
+  this.src = message.src;
+  this._pending = message;
+  this._setPendingPresence();
+};
+
+RemoteDocPresence.prototype.destroy = function(callback) {
+  this._doc.removeListener('op', this._opHandler);
+  this._doc.removeListener('create', this._createDelHandler);
+  this._doc.removeListener('del', this._createDelHandler);
+  this._doc.removeListener('load', this._loadHandler);
+
+  RemotePresence.prototype.destroy.call(this, callback);
+};
+
+RemoteDocPresence.prototype._registerWithDoc = function() {
+  this._doc.on('op', this._opHandler);
+  this._doc.on('create', this._createDelHandler);
+  this._doc.on('del', this._createDelHandler);
+  this._doc.on('load', this._loadHandler);
+};
+
+RemoteDocPresence.prototype._setPendingPresence = function() {
+  if (this._pendingSetPending) return;
+  this._pendingSetPending = true;
+  var presence = this;
+  this._doc.whenNothingPending(function() {
+    presence._pendingSetPending = false;
+    if (!presence._pending) return;
+    if (presence._pending.seq < presence.seq) return presence._pending = null;
```

Comment:
Since a client connection's `seq` currently resets to 1 (or 0?) upon reconnecting, I'm worried about buggy behavior in scenario:
1. Client A does some ops and presence pushing, ending up at a `seq` of 20.
2. Client B is subscribed to Client A's presence, so its RemotePresence for Client A has a `presence.seq` of 20.
3. Client A disconnects and reconnects, so its `seq` resets. It starts sending presence again with `seq` <20.
4. Client B ignores those messages from Client A until the `seq` is over 20 again.

As we discussed, without the assumption of persistent src+seq for a client session, we'll either have to take this out, or to add a presence version.

If we go the presence versioning route, we'll similarly want to be cautious about two sessions using the same `presenceId`, otherwise two sessions with the same `presenceId` can overwrite each others' presence.

Reply:
I've addressed the `seq` issue by just adding a `pv` presence version field to the message instead. (NB: subscribe / unsubscribe still uses a `seq`, but that's derived in `Presence`, separately from the `Connection` `seq` so should be fine to use. Maybe could do with a better name...).

As far as `presenceId` being unique goes, I think I'm not feeling very comfortable with suffixing the ID. I think having two different IDs for something is quite confusing, both in the codebase (I've already tripped over the channel namespacing a few times, for example), and from the consumer's point of view (eg when directly accessing a local presence by ID in `presence.localPresences`).

A possible middle ground could be providing a sensible default random ID if the presence ID is not provided (ie make the ID optional), so that we get sensible default behaviour, and if the client doesn't care about what the presence ID represents, then they don't have to care (they can naturally still query `localPresence.presenceId` to find out what random ID has been assigned to them).